More container fields for SuggestionConfig

fischor opened the pull request:
<!--  Thanks for sending a pull request! Here are some tips for you:
1. If this is your first time, check our contributor guidelines https://www.kubeflow.org/docs/about/contributing
2. To know more about Katib components, check developer guide https://github.com/kubeflow/katib/blob/master/docs/developer-guide.md
3. If you want *faster* PR reviews, check how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
-->

**What this PR does / why we need it**:

This PR adds more fields to the `SuggestionConfig` that is used to populate the main container of the Suggestion deployment. It adds fields `command`, `args`, `envs` that can be used to pass arguments to the executable. This allows the user to easily parametrize the suggestion algorithm/executable itself by passing values using these new fields.   

**Which issue(s) this PR fixes**:
Fixes #1737 (at least partly)

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/): the docs will need to mention the new fields

andreyvelich commented on pkg/util/v1beta1/katibconfig/config.go:
```diff
@@ -36,6 +36,11 @@ import (
+	Args                      []string                         `json:"args,omitempty"`
+	WorkingDir                string                           `json:"workingDir,omitempty"`
+	EnvFrom                   []corev1.EnvFromSource           `json:"envFrom,omitempty"`
+	Env                       []corev1.EnvVar                  `json:"env,omitempty"`
```
Maybe we could start adding more fields to SuggestionConfig with `corev1.Container{}` field, similar to [`PodSpec` API in K8s](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3105) ? I think you can use inline settings to avoid modification in existing Katib config YAML:
```go
 v1.Container `json:",inline"`
 ```

I guess, at the end we want to give user ability to modify the whole Suggestion's Deployment spec via Katib config, but it requires much more changes.

WDYT @tenzen-y @johnugeorge @fischor ?

tenzen-y replied:
> Maybe we could start adding more fields to SuggestionConfig with corev1.Container{} field, similar to [PodSpec API in K8s](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3105) ? I think you can use inline param to avoid modification in existing Katib config YAML

sgtm

> I guess, at the end we want to give user ability to modify the whole Suggestion's Deployment spec via Katib config, but it requires much more changes.

I think so too. Maybe, to support customizing Tolerations, imagePullSecret and so on, we want to support customizing PodSpec for suggestion service deployment.

fischor replied:
Yes, inlining `corev1.Container` should also be doable.  
Certain fields of the container will be set by the katib-Controller, like some ports, the liveness and readiness probes and possibly the "suggestion-volume". How to handle these?

> I think so too. Maybe, to support customizing Tolerations, imagePullSecret and so on, we want to support customizing PodSpec for suggestion service deployment.

To support his, how about specifying a default deployment in the suggestion config (instead of just the container).

johnugeorge replied:
For namespace specific specific suggestion deployments, we can allow to have a configmap in a user namespace(where experiment is run). If suggestion spec is provided in the configmap of the user namespace, it can override the global katib config.  
Ref: https://github.com/kubeflow/katib/issues/1976

andreyvelich replied:
> Certain fields of the container will be set by the katib-Controller, like some ports, the liveness and readiness probes and possibly the "suggestion-volume". How to handle these? Overwrite or not?

@fischor I think if parameters are set by the user we should use them, otherwise use the default values.

> To support his, how about specifying a default deployment in the suggestion config (instead of just the container)?.

We might want to set `spec.template.spec` for our Suggestion config, as I mentioned here: https://github.com/kubeflow/katib/issues/1737#issuecomment-974245081, but that requires much more changes since we need to modify manifests.

>  If suggestion spec is provided in the configmap of the user namespace, it can override the global katib config.

@johnugeorge That might be the next step. We need to have discussion what is the best way to give user ability to set experiment-scoped Katib Config.

andreyvelich reviewed (commented):
Thank you for taking this @fischor!

johnugeorge commented:
@fischor Can you rebase and fix?

fischor commented:
The name of the suggestion algorithm container was not defaulted. Creating the suggestion deployment failed because of a missing container  name.

@johnugeorge I have rebased and fixed.

andreyvelich commented on pkg/util/v1beta1/katibconfig/config.go:
```diff
@@ -34,9 +34,7 @@ import (
-	Resource                  corev1.ResourceRequirements      `json:"resources,omitempty"`
+	corev1.Container          `json:",inline"`
 	ServiceAccountName        string                           `json:"serviceAccountName,omitempty"`
 	VolumeMountPath           string                           `json:"volumeMountPath,omitempty"`
```
@fischor Do we need `VolumeMountPath` if that is part of Container parameters: https://github.com/kubeflow/katib/blob/7f1b2842f87158c25c54f577da1b82ddc7a958e3/pkg/controller.v1beta1/suggestion/composer/composer.go#L237 ?

fischor replied:
The current Katib `SuggestionConfig` has a JSON field `volumeMountPath` thats just a filepath, the `corev1.Container` has a json field `volumeMounts` thats an array of volume mounts. So `volumeMountPath` is not part if the container fields.

The referenced snippet kind of follows the logic:
> ... if parameters are set by the user we should use them, otherwise use the default values.

In case the `volumeMounts` are set we prefer that, otherwise mount the default suggestion volume under the specificed `volumeMountPath`.

andreyvelich replied:
I see, that is because we set [the default value name for Suggestion Volume](https://github.com/kubeflow/katib/blob/7f1b2842f87158c25c54f577da1b82ddc7a958e3/pkg/controller.v1beta1/suggestion/composer/composer.go#L121). Please can you leave comment that we should refactor `VolumeMountPath` parameter once we allow to set Deployment spec in our Suggestion config ?

Also, I think we should modify [`len(suggestionContainer.VolumeMounts) == 0` check](https://github.com/kubeflow/katib/blob/7f1b2842f87158c25c54f577da1b82ddc7a958e3/pkg/controller.v1beta1/suggestion/composer/composer.go#L237).
We always should append `VolumeMount` if ResumePolicy is `FromVolume`.

```go
if s.Spec.ResumePolicy == experimentsv1beta1.FromVolume {
volumeMount := corev1.VolumeMount{
	Name:      consts.ContainerSuggestionVolumeName,
	MountPath: suggestionConfigData.VolumeMountPath,
}
suggestionContainer.VolumeMounts = append(suggestionContainer.VolumeMounts, volumeMount)
}
```
WDYT @fischor ?

fischor replied:
@andreyvelich Have you seen that the consts.DefaultSuggestionPortName port is also just set if the Container specified in the suggestion has no ports set? See https://github.com/fischor/katib/blob/7f1b2842f87158c25c54f577da1b82ddc7a958e3/pkg/controller.v1beta1/suggestion/composer/composer.go#L194-L201
Would you also change that to always append the default port?

I think both ways of doing this have pros/cons. If we always append the default volume/port then the user has no option to overwrite it. If we never append the default volume/port, then the user needs to be aware that it has to specify it when setting any volumes/ports.

A third option would be to check if the default volume/port is already set in the container and only append if its not. I think I would prefer that.

andreyvelich replied:
> A third option would be to check if the default volume/port is already set in the container and only append if its not. I think I would prefer that.

I think that should work, but we need to make sure that volume and port name is correct (`DefaultSuggestionPortName` and `ContainerSuggestionVolumeName`).

andreyvelich commented on pkg/controller.v1beta1/suggestion/composer/composer.go:
```diff
@@ -181,21 +181,30 @@ func (g *General) desiredContainers(s *suggestionsv1beta1.Suggestion,
 	}
 
-	if viper.GetBool(consts.ConfigEnableGRPCProbeInSuggestion) {
+	if !containsContainerPortWithName(suggestionContainer.Ports, consts.DefaultSuggestionPortName) &&
+		!containsContainerPort(suggestionConfigData.Ports, consts.DefaultSuggestionPort) {
```
Should we verify that port `name` and `containerPort` is set together in the Katib Config ?
[`DefaultSuggestionPortName` and `DefaultSuggestionPort` is always used for the Suggestion's service](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/suggestion/composer/composer.go#L147-L151).
So if user set `DefaultSuggestionPortName` port with different address, the Suggestion service failed.

cc @tenzen-y

andreyvelich replied:
cc @johnugeorge

tenzen-y replied:
It might be better to prohibit setting containerPort in katibConfig.

andreyvelich replied:
@fischor Do you have time to check my latest commit ? We are getting closer for the Katib 0.15 release, and it would be great to include your feature in that release.

fischor replied:
@andreyvelich I should have some spare time tomorrow. Do we do it as you suggested: verify that name and port are correct, otherwise return an error in `DesiredDeployment` method? Or prohibit the DefaultSuggestionPort to be set at all?

andreyvelich replied:
@fischor Thank you! Let's prohibit to set `DefaultSuggestionPort` for now as @tenzen-y suggested. So we can avoid errors from the user side.

johnugeorge replied:
@fischor We are planning to cut a release in a day. Do you have any update?

fischor replied:
@johnugeorge Thanks for the reminder...you can check the latest commit now.

andreyvelich commented:
Thanks you for updates @fischor.
Excited to see a new usage of Katib config!
/lgtm
/approve

tenzen-y commented:
@fischor Great work!
Thanks for your effort!
